README.md
# GitHub Action to Add Release Metadata to The IOTA Tangle

This GitHub Action will take the contents of your GitHub tagged release and create an associated transaction on the IOTA Tangle. This way the data associated with the release becomes immutable.

The action will perform the following steps:

* Load the release tagged by the input `tag_name`
* Create a signature for the content of `tarball_url`
* Create a signature for the content of `zipball_url`
* Create a signature for each of the attached assets
* Attach a payload to the IOTA Tangle with information about the release and the hashes

### Inputs

- `tag_name`: The name of the tag for this release
- `comment`: An optional comment to include in the Tangle payload

### Outputs

- `tx_hash`: The hash of the transaction on the Tangle
- `tx_explore_url`: A url which can be used to explore the transaction on the Tangle


## Example Payload

```json
{
   "owner": "an-owner",
   "repo": "test-repo",
   "tag_name": "v0.1.0",
   "name": "First full release",
   "comment": "My Awesome Release",
   "body": "This is the body of the description.",
   "tarball_url": "https://api.github.com/repos/an-owner/test-repo/tarball/v0.1.0",
   "tarball_sig": "Me3ouGni0h50TOHQklopu3sJdLFh/ZVlPJom3aDRFVQ=",
   "zipball_url": "https://api.github.com/repos/an-owner/test-repo/zipball/v0.1.0",
   "zipball_sig": "jyQ8U1T4oMSEbT3e9NTuFyoMskwAvti3nmiYKtuh8LU=",
   "assets": [
      {
         "name": "attach-1.zip",
         "size": 150752,
         "url": "https://github.com/an-owner/test-repo/releases/download/v0.1.0/attach-1.zip",
         "sig": "YQylonV2i+5KtwVN0FxTU7ssWflX+6fC29COSbFOmfQ="
      },
      {
         "name": "attach-2.zip",
         "size": 153054,
         "url": "https://github.com/an-owner/test-repo/releases/download/v0.1.0/attach-2.zip",
         "sig": "a+Rgpf5gs0lpCJ8wt+eymkTdo99RbcP0o1PgLCIT2NE="
      }
   ]
}
```

## Usage

Create a GitHub workflow in you repo e.g. `/.github/workflows/tangle-release.yml`.
 Most of the environment variables are optional, except for the `IOTA_SEED` which must be 81 trytes in length. For more details on creating a seed see [IOTA Docs - Getting Started - Creating A Seed](https://docs.iota.org/docs/getting-started/0.1/tutorials/create-a-seed)

```yaml
on:
  push:
    tags:
      - 'v*' # Push events to matching v*

name: Create Immutable Release

jobs:
  build:
    name: Create Immutable Release
    runs-on: ubuntu-latest
    steps:
      - name: Tangle Release
        id: tangle_release
        uses: iotaledger/gh-tangle-release@v0.7.2
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          IOTA_SEED: ${{ secrets.IOTA_SEED }}
          IOTA_NODE: ${{ secrets.IOTA_NODE }} # Optional, defaults to https://nodes.iota.cafe:443
          IOTA_ADDRESS_INDEX: ${{ secrets.IOTA_ADDRESS_INDEX }} # Optional, defaults to 0
          IOTA_DEPTH: ${{ secrets.IOTA_DEPTH }} # Optional, defaults to 3
          IOTA_MWM: ${{ secrets.IOTA_MWM }} # Optional, defaults to 14
          IOTA_TAG: ${{ secrets.IOTA_TAG }} # Optional, defaults to GITHUB9RELEASE
          IOTA_EXPLORE_URL: ${{ secrets.IOTA_EXPLORE_URL }} # Optional, defaults to https://utils.iota.org/transaction/:hash
        with:
          tag_name: ${{ github.ref }}
          comment: My Awesome Release

```

## CLI

The core features are also available to use as a CLI independently of GitHub Actions.

If you install the npm package with:

```shell
npm install @iota/gh-tangle-release
```

You can then execute the following script.

```shell
gh-tangle-release
```

You will then be presented with the following options.

```shell
GitHub Tangle Release v0.7.2 🚀

Usage: gh-tangle-release [options]

An application for creating a transaction on the IOTA Tangle from a GitHub release.

Options:
  -v, --version               output the current version
  --github-token <string>     GitHub token for accessing your repository (required)
  --owner <string>            GitHub repository owner (required)
  --repository <string>       GitHub repository (required)
  --release-tag <string>      The release tag from the GitHub repository (required)
  --node <string>             Url of the node to use for attaching the transaction to the tangle (default:
                              "https://nodes.iota.cafe:443")
  --depth <number>            Depth to use for attaching the transaction to the tangle (default: "3")
  --mwm <number>              Minimum weight magnitude to use for attaching the transaction to the tangle (default: "14")
  --seed <string>             81 Tryte seed used to generate addresses (required)
  --address-index <number>    Index number used to generate addresses (default: "0")
  --transaction-tag <string>  Tag to apply to the Tangle transaction (default: "GITHUB9RELEASE")
  --comment <string>          An optional comment to include in the Tangle transaction payload
  --explorer-url <string>     Url of the explorer to use for exploration link (default: "https://utils.iota.org/transaction/:hash")
  --no-color                  Disable colored output
  --help                      Display help

You can also supply some of the options through environment variables:
   --github-token: GITHUB_TOKEN
   --owner: GITHUB_REPOSITORY[0]
   --repository: GITHUB_REPOSITORY[1]
        where GITHUB_REPOSITORY is formatted owner/repository
   --release-tag: GITHUB_REF
   --seed: GTR_SEED


Example: gh-tangle-release --github-token a4d936470cb3d66f5434f787c2500bde9764f --owner my-org --repository my-repo --release-tag v1.0.1 --seed AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
  ```


CHANGELOG.md
# Changelog

## v0.7.2

Removed superfluous console.log

## v0.7.1

Show additional defaults in CLI help
Fix missing address index in cli
Added env var processing for GITHUB_TOKEN, GITHUB_REPOSITORY, GITHUB_REF and GTR_SEED
As we now support env vars for the required options it is valid to have no command line parameters
All parameter validation errors shown at once
Added --no-color option

## v0.7.0

Refactored to include CLI
Rewritten in TypeScript

## v0.6.2

Working release

src/cli-core.ts
import chalk from "chalk";
import { Command } from "commander";
import emoji from "node-emoji";
import { sanitizeInput, tangleRelease } from "./core";

/**
 * Execute the cli core.
 * @param argv The command line arguments.
 * @param env The environment variables.
 * @param display Method to output display.
 */
export async function cliCore(
    argv: string[],
    env: { [id: string]: string | undefined },
    display: (message: string, param?: string) => void): Promise<void> {
    const program = new Command();

    try {
        const version = "0.7.2";

        program
            .storeOptionsAsProperties(false)
            .passCommandToAction(false)
            .name(chalk.yellowBright("gh-tangle-release"))
            .version(version, "-v, --version", chalk.yellowBright("output the current version"))
            .description(
                chalk.cyan("An application for creating a transaction on the IOTA Tangle from a GitHub release."))
            .option("--github-token <string>", chalk.yellowBright(
                "GitHub token for accessing your repository (required)"))
            .option("--owner <string>", chalk.yellowBright("GitHub repository owner (required)"))
            .option("--repository <string>", chalk.yellowBright("GitHub repository (required)"))
            .option("--release-tag <string>",
                chalk.yellowBright("The release tag from the GitHub repository (required)"))
            .option("--node <string>",
                chalk.yellowBright("Url of the node to use for attaching the transaction to the tangle"),
                "https://nodes.iota.cafe:443")
            .option("--depth <number>", chalk.yellowBright("Depth to use for attaching the transaction to the tangle"),
                "3")
            .option("--mwm <number>",
                chalk.yellowBright("Minimum weight magnitude to use for attaching the transaction to the tangle"),
                "14")
            .option("--seed <string>", chalk.yellowBright("81 Tryte seed used to generate addresses (required)"))
            .option("--address-index <number>", chalk.yellowBright("Index number used to generate addresses"), "0")
            .option("--transaction-tag <string>", chalk.yellowBright("Tag to apply to the Tangle transaction"),
                "GITHUB9RELEASE")
            .option("--comment <string>",
                chalk.yellowBright("An optional comment to include in the Tangle transaction payload"))
            .option("--explorer-url <string>", chalk.yellowBright("Url of the explorer to use for exploration link"),
                "https://utils.iota.org/transaction/:hash")
            .option("--no-color", chalk.yellowBright("Disable colored output"))
            .helpOption("--help",
                chalk.yellowBright("Display help"));

        program.parse(argv);
        const opts = program.opts();

        display(chalk.green(`GitHub Tangle Release v${version} ${
            opts.color === false ? "" : emoji.get("rocket")}\n`));

        const envRepo: string[] = env.GITHUB_REPOSITORY ? env.GITHUB_REPOSITORY.split("/") : [];

        if (envRepo.length === 2) {
            opts.owner = opts.owner || envRepo[0];
            opts.repository = opts.repository || envRepo[1];
        }

        const config = sanitizeInput({
            githubToken: opts.githubToken || env.GITHUB_TOKEN,
            owner: opts.owner,
            repository: opts.repository,
            releaseTag: opts.releaseTag || env.GITHUB_REF,
            node: opts.node,
            depth: opts.depth,
            mwm: opts.mwm,
            seed: opts.seed || env.GTR_SEED,
            addressIndex: opts.addressIndex,
            transactionTag: opts.transactionTag,
            comment: opts.comment,
            explorerUrl: opts.explorerUrl
        });

        display("Options:");
        display(chalk.cyan("\tGitHub Token"), chalk.white("*******"));
        display(chalk.cyan("\tOwner"), chalk.white(config.owner));
        display(chalk.cyan("\tRepository"), chalk.white(config.repository));
        display(chalk.cyan("\tRelease Tag"), chalk.white(config.releaseTag));
        display(chalk.cyan("\tNode"), chalk.white(config.node));
        display(chalk.cyan("\tDepth"), chalk.white(config.depth));
        display(chalk.cyan("\tMWM"), chalk.white(config.mwm));
        display(chalk.cyan("\tSeed"), chalk.white("*******"));
        display(chalk.cyan("\tAddress Index"), chalk.white(config.addressIndex));
        display(chalk.cyan("\tTransaction Tag"), chalk.white(config.transactionTag));
        if (config.comment) {
            display(chalk.cyan("\tComment"), chalk.white(config.comment));
        }
        display(chalk.cyan("\tExplorer Url"), chalk.white(config.explorerUrl));
        display("");

        try {
            const transactionDetails = await tangleRelease(config, message => display(chalk.green(message)));

            display("Transaction Hash:",
                chalk.cyan(transactionDetails.hash));
            display("You can view the transaction on the tangle at:",
                chalk.cyan(transactionDetails.url));
            display(chalk.green("Complete"));
        } catch (err) {
            display("");
            display(createErrors(err));
            process.exit(1);
        }
    } catch (err) {
        program.help(str => `${str}${createEnvHelp()}${createExample()}${createErrors(err)}`);
        process.exit(1);
    }
}

/**
 * Show an example on the console.
 * @returns The example text.
 */
function createExample(): string {
    // eslint-disable-next-line max-len
    return chalk.magenta("\nExample: gh-tangle-release --github-token a4d936470cb3d66f5434f787c2500bde9764f --owner my-org --repository my-repo --release-tag v1.0.1 --seed AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n\n");
}

/**
 * Show additional info about env vars.
 * @returns The additional information.
 */
function createEnvHelp(): string {
    return `
${chalk.cyan("You can also supply some of the options through environment variables:")}
   ${chalk.cyan("--github-token: GITHUB_TOKEN")}
   ${chalk.cyan("--owner: GITHUB_REPOSITORY[0]")}
   ${chalk.cyan("--repository: GITHUB_REPOSITORY[1]")}
   ${chalk.cyan("     where GITHUB_REPOSITORY is formatted owner/repository")}
   ${chalk.cyan("--release-tag: GITHUB_REF")}
   ${chalk.cyan("--seed: GTR_SEED")}\n\n`;
}

/**
 * Show the errors.
 * @param error The error that was thrown.
 * @returns The formatted errors.
 */
function createErrors(error: Error): string {
    return chalk.red(`The following errors occurred:\n   ${error.message.replace(/\n/g, "\n   ")}`);
}
